### PR TITLE
Feature/debugging 2028

### PR DESCRIPTION
**Pull request by Daboo1998**

### CODE REVIEW

File: src/scenes/collaboration/Collaboration.tsx [hunk 1]
```diff
@@ -251,32 +241,13 @@ const Collaboration: FC = () => {
...
-        setNewDevice(devices[1]);
-        newDeviceDialog.open();
-      });
-    };
-
-    return () => {
-      navigator.mediaDevices.ondevicechange = null;
-    };
+  useDeviceChange(() => {
+    navigator.mediaDevices.enumerateDevices().then((devices) => {
```
Comment (PAKovalev) [inline src/scenes/collaboration/Collaboration.tsx:245]: move `navigator` and `newDevice useState` to hook also.

File: src/shared/hooks/useDeviceChange.tsx [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+export const useDeviceChange = (callback: () => void, deps: DependencyList = []) => {
+  useEffect(() => {
+    navigator.mediaDevices.ondevicechange = callback;
+
+    return () => {
+      navigator.mediaDevices.ondevicechange = null;
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [callback, ...deps]);
```
Comment (PAKovalev) [inline src/shared/hooks/useDeviceChange.tsx:11]: hmm... strange

**Review by PAKovalev — CHANGES_REQUESTED** [review]

File: src/stores/MainStore/models/AgoraStore/AgoraScreenShareStore/AgoraScreenShareStore.ts [hunk 3]
```diff
@@ -36,20 +40,64 @@ const AgoraScreenShareStore = types
...
+  }))
+  // Listeners registration
+  .actions((self) => ({
+    setupAgoraListeners() {
+      self.client?.on('user-published', self.handleUserPublished);
+      self.client?.on('user-unpublished', self.handleUserPublished);
+      self.client?.on('user-left', self.handleUserLeft);
+      self.client?.on('connection-state-change', self.handleConnectionStateChange);
+    },
+    clanupListeners() {
```
Comment (aminbksi) [inline src/stores/MainStore/models/AgoraStore/AgoraScreenShareStore/AgoraScreenShareStore.ts:99]: cleanup you mean?

File: src/stores/MainStore/models/AgoraStore/AgoraStageModeStore/AgoraStageModeStore.ts [hunk 4]
```diff
@@ -130,6 +146,115 @@ const AgoraStageModeStore = types
...
+  .actions((self) => ({
+    setupAgoraListeners() {
+      self.client.on('user-published', self.handleUserPublished);
+      self.client.on('user-unpublished', self.handleUserPublished);
+      self.client.on('user-joined', self.handleUserJoined);
+      self.client.on('user-left', self.handleUserLeft);
+      self.client.on('connection-state-change', self.handleConnectionStateChange);
+      self.client.on('volume-indicator', self.handleVolumeIndicator);
+    },
+    clanupListeners() {
```
Comment (aminbksi) [inline src/stores/MainStore/models/AgoraStore/AgoraStageModeStore/AgoraStageModeStore.ts:254]: same

**Review by PAKovalev — APPROVED** [review]

**Review by aminbksi — APPROVED** [review]

**Pull request merged**